src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs

```diff
@@ -94,28 +145,53 @@ protected internal XmlDocument(XmlImplementation imp) : base()
             _implementation = imp;
             _domNameTable = new DomNameTable(this);
 
-            // force the following string instances to be default in the nametable
-            XmlNameTable nt = this.NameTable;
-            nt.Add(string.Empty);
-            strDocumentName = nt.Add("#document");
-            strDocumentFragmentName = nt.Add("#document-fragment");
-            strCommentName = nt.Add("#comment");
-            strTextName = nt.Add("#text");
-            strCDataSectionName = nt.Add("#cdata-section");
-            strEntityName = nt.Add("#entity");
-            strID = nt.Add("id");
-            strNonSignificantWhitespaceName = nt.Add("#whitespace");
-            strSignificantWhitespaceName = nt.Add("#significant-whitespace");
-            strXmlns = nt.Add("xmlns");
-            strXml = nt.Add("xml");
-            strSpace = nt.Add("space");
-            strLang = nt.Add("lang");
-            strReservedXmlns = nt.Add(XmlReservedNs.NsXmlNs);
-            strReservedXml = nt.Add(XmlReservedNs.NsXml);
-            strEmpty = nt.Add(String.Empty);
+            strXmlns = Xmlns;
+            strXml = Xml;
+            strReservedXmlns = XmlReservedNs.NsXmlNs;
+            strReservedXml = XmlReservedNs.NsXml;
             baseURI = String.Empty;
-
             objLock = new object();
+
+            if (imp.NameTable.GetType() == typeof(NameTable))
+            {
+                // When the name table being used is of type NameTable avoid re-calculating the hash codes.
+                NameTable nt = (NameTable)imp.NameTable;
+
+                strDocumentName = nt.GetOrAddEntry(s_nameTableSeeds[DocumentNameSeedIndex].key, s_nameTableSeeds[DocumentNameSeedIndex].hash);
```

Comment:
is it worth defining the constant integers? `    strDocumentName = nt.GetOrAddEntry(s_nameTableSeeds[1].key, s_nameTableSeeds[1].hash);` seems quite maintainable

Reply:
I had it both ways (1st without the constants), the only thing is that with the constant is easier to write a regex to help verify if most of them are correct, other than that it is a wash. I don't have much of a preference here, leaning to keep as it is, but there is no hurry to push this PR so I will give some time for you guys to consider...

Reply:
The indexes are constants the str* fields is because the whole purpose of the table is to allow reference comparison instead of string comparison. In an earlier version of this PR I was removing the fields but we can't because if the table already has them (a corner case granted) we need the fields to use the same as the ones already in the table. See https://github.com/dotnet/corefx/pull/29626#discussion_r187467176